### PR TITLE
[Enhancement] Jsonpaths does not needed for avro union data

### PR DESCRIPTION
**Pull request by zaorangyang**
## Problem Summary:
For avro data that does not have nested data, the user must specify the jsonpaths when parsing union data. If you're loading a table with hundreds or thousands of columns, you need to write jsonpath for each column, which is  tedious. This PR fixes this issue.

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/avro_scanner.cpp [hunk 1]
```diff
@@ -362,8 +362,15 @@ Status AvroScanner::_construct_row_without_jsonpath(const avro_value_t& avro_val
...
+        // We should expand the union type.
+        avro_value_t& cur_value = element_value;
+        if (UNLIKELY(avro_value_get_type(&cur_value) == AVRO_UNION)) {
+            avro_value_t branch;
+            RETURN_IF_ERROR(_handle_union(cur_value, branch));
+            cur_value = branch;
+        }
         // construct column with value.
-        RETURN_IF_ERROR(_construct_column(element_value, column.get(), slot_info.type_, slot_info.key_));
+        RETURN_IF_ERROR(_construct_column(cur_value, column.get(), slot_info.type_, slot_info.key_));
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:373]: The name format in StarRocks is `_key` not `key_`. The `_` is a prefix in member name, not a suffix.

File: be/src/exec/avro_scanner.cpp [hunk 2]
```diff
@@ -362,8 +362,15 @@ Status AvroScanner::_construct_row_without_jsonpath(const avro_value_t& avro_val
         }
 
         auto& column = chunk->get_column_by_slot_id(slot_info.id_);
+        // We should expand the union type.
+        avro_value_t& cur_value = element_value;
+        if (UNLIKELY(avro_value_get_type(&cur_value) == AVRO_UNION)) {
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:367]: What's the usage of `AVRO_UNION`?

File: be/src/exec/avro_scanner.cpp [hunk 3]
```diff
@@ -362,8 +362,15 @@ Status AvroScanner::_construct_row_without_jsonpath(const avro_value_t& avro_val
...
+            avro_value_t branch;
+            RETURN_IF_ERROR(_handle_union(cur_value, branch));
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:369]: We will pass a pointer to a function to let it be changed. If you pass a reference, it's better to define it constant like `const avro_value_t&`

File: be/src/exec/avro_scanner.cpp (hunk 2 above)
Comment (zaorangyang) [reply be/src/exec/avro_scanner.cpp:367]: > What's the usage of `AVRO_UNION`?

nullable data type

File: be/src/exec/avro_scanner.cpp (hunk 1 above)
Comment (zaorangyang) [reply be/src/exec/avro_scanner.cpp:373]: > The name format in StarRocks is `_key` not `key_`. The `_` is a prefix in member name, not a suffix.
I will create a new pr to solve this issue.

File: be/src/exec/avro_scanner.cpp (hunk 3 above)
Comment (zaorangyang) [reply be/src/exec/avro_scanner.cpp]: > We will pass a pointer to a function to let it be changed. If you pass a reference, it's better to define it constant like `const avro_value_t&`

updated

**Review by chaoyli — DISMISSED** [review]

**Review by meegoo — DISMISSED** [review]

**Review by meegoo — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**